File: cc_pomodoro/state.py
from __future__ import annotations

import json
import uuid
from datetime import datetime, timedelta, timezone
from typing import Any

from cc_pomodoro.constants import STATE_FILE
from cc_pomodoro.config import _atomic_write, _ensure_config_dir


def _default_state() -> dict[str, Any]:
    return {
        "active": False,
        "session_id": None,
        "started_at": None,
        "end_at": None,
        "duration": 0,
        "app": None,
    }


def get_state() -> dict[str, Any]:
    if STATE_FILE.exists():
        raw = STATE_FILE.read_text(encoding="utf-8")
        if raw.strip():
            stored = json.loads(raw)
            return stored
    return dict(_default_state())


def is_active() -> bool:
    state = get_state()
    if not state.get("active"):
        return False
    remaining = get_remaining_seconds()
    if remaining <= 0:
        return False
    return True


def get_remaining_seconds() -> int:
    state = get_state()
    if not state.get("active"):
        return 0
    end_at_str = state.get("end_at")
    if not end_at_str:
        return 0
    try:
        end_at = datetime.fromisoformat(end_at_str)
        now = datetime.now(timezone.utc)
        remaining = (end_at - now).total_seconds()
        return max(0, int(remaining))
    except (ValueError, TypeError):
        return 0


def get_session_id() -> str | None:
    state = get_state()
    return state.get("session_id")


def start_session(duration_min: int, app: str) -> str:
    session_id = str(uuid.uuid4())
    now = datetime.now(timezone.utc)
    started_at = now.isoformat()
    end_at = (now + timedelta(minutes=duration_min)).isoformat()
    state = {
        "active": True,
        "session_id": session_id,
        "started_at": started_at,
        "end_at": end_at,
        "duration": duration_min,
        "app": app,
    }
    _ensure_config_dir()
    _atomic_write(STATE_FILE, json.dumps(state, indent=2, ensure_ascii=False) + "\n")
    return session_id


def end_session() -> dict[str, Any]:
    state = get_state()
    state["active"] = False
    _ensure_config_dir()
    _atomic_write(STATE_FILE, json.dumps(state, indent=2, ensure_ascii=False) + "\n")
    return state

File: src/state.ts
import crypto from 'node:crypto';
import fs from 'node:fs';
import { STATE_FILE } from './constants.js';
import { atomicWrite, ensureConfigDir } from './config.js';

export interface State {
  active: boolean;
  session_id: string | null;
  started_at: string | null;
  end_at: string | null;
  duration: number;
  app: string | null;
  [key: string]: unknown;
}

function defaultState(): State {
  return {
    active: false,
    session_id: null,
    started_at: null,
    end_at: null,
    duration: 0,
    app: null,
  };
}

/**
 * Read the current state from disk.
 */
export function getState(): State {
  if (fs.existsSync(STATE_FILE)) {
    const raw = fs.readFileSync(STATE_FILE, 'utf-8').trim();
    if (raw) {
      return JSON.parse(raw) as State;
    }
  }
  return defaultState();
}

/**
 * Return true if a session is active and has not expired.
 */
export function isActive(): boolean {
  const state = getState();
  if (!state.active) return false;
  const remaining = getRemainingSeconds();
  return remaining > 0;
}

/**
 * Return remaining seconds for the active session (0 if no active session).
 */
export function getRemainingSeconds(): number {
  const state = getState();
  if (!state.active || !state.end_at) return 0;
  try {
    const endAt = new Date(state.end_at).getTime();
    const now = Date.now();
    return Math.max(0, Math.floor((endAt - now) / 1000));
  } catch {
    return 0;
  }
}

/**
 * Get the current session ID, or null if no session.
 */
export function getSessionId(): string | null {
  return getState().session_id;
}

/**
 * Start a new session, writing state to disk.
 * Returns the generated session UUID.
 */
export function startSession(durationMin: number, app: string): string {
  const sessionId = crypto.randomUUID();
  const now = new Date();
  const startedAt = now.toISOString();
  const endAt = new Date(now.getTime() + durationMin * 60 * 1000).toISOString();
  const state: State = {
    active: true,
    session_id: sessionId,
    started_at: startedAt,
    end_at: endAt,
    duration: durationMin,
    app,
  };
  ensureConfigDir();
  atomicWrite(STATE_FILE, JSON.stringify(state, null, 2) + '\n');
  return sessionId;
}

/**
 * End the current session (mark active=false).
 */
export function endSession(): State {
  const state = getState();
  state.active = false;
  ensureConfigDir();
  atomicWrite(STATE_FILE, JSON.stringify(state, null, 2) + '\n');
  return state;
}
